src/vos/tests/vts_ilog.c

```diff
@@ -554,16 +547,26 @@ ilog_test_update(void **state)
 	rc = entries_check(umm, ilog, &ilog_callbacks, NULL, 0, entries);
 	assert_int_equal(rc, 0);
 
-	/** Same epoch, different DTX */
+	/** Same epoch, different transaction, same operation.  In other
+	 *  words, both the existing entry and this one are punches so
+	 *  we get back -DER_ALREADY because the existing entry covers
+	 *  this punch.
+	 */
 	fake_tx_reset();
-	rc = ilog_update(loh, NULL, epoch, true);
-	if (rc != -DER_TX_RESTART) {
-		print_message("Epoch entry already exists.  Replacing with"
-			      " different DTX should get -DER_TX_RESTART:"
-			      " rc=%s\n",
-			      d_errstr(rc));
-		assert(0);
-	}
+	rc = ilog_update(loh, NULL, epoch, 3, true);
+	LOG_FAIL(rc, -DER_ALREADY, "Epoch entry already exists. "
+		 "Replacing with different DTX should get "
+		 "-DER_ALREADY\n");
+
+	/** Same epoch, different DTX, different operation operation.
+	 *  Trying to replace a punch with an update at the same
+	 *  epoch requires a restart with later epoch
+	 */
```

Comment:
Then how about replace an "update" with a "punch" at the same epoch, also "-DER_TX_RESTART", right? But the ilog_update() function parameter only can know whether this one is "punch" or not, it does not know the existing one is "punch" or "update", then the "different operation" seems not clear.

Reply:
The in tree one is assumed a punch or update based on the current value of the minor epochs